### PR TITLE
SI-8283 mutation-free bound inference for existentials

### PR DESCRIPTION
**Pull request by adriaanm**
A safer version of the fix for SI-6169 (#3471)

review by @retronym

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 1]
```diff
@@ -384,9 +384,14 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
      */
-    def newExistentialSkolem(basis: Symbol, origin: AnyRef): TypeSkolem = {
-      val skolem = newTypeSkolemSymbol(basis.name.toTypeName, origin, basis.pos, (basis.flags | EXISTENTIAL) & ~PARAM)
-      skolem setInfo (basis.info cloneInfo skolem)
+    def newExistentialSkolem(basis: Symbol, origin: AnyRef): TypeSkolem =
+      newExistentialSkolem(basis.name.toTypeName, basis.info, basis.flags, basis.pos, origin)
+
+    /** Create a new existential type skolem with this symbol its owner, and the given other properties.
+     */
+    def newExistentialSkolem(name: TypeName, info: Type, flags: Long, pos: Position, origin: AnyRef): TypeSkolem = {
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala:392]: Ha, you read my mind.

File: src/reflect/scala/reflect/internal/Types.scala [hunk 2]
```diff
@@ -2564,43 +2568,39 @@ trait Types
...
-    override def skolemizeExistential(owner: Symbol, origin: AnyRef) = {
       // do this here because it's quite close to what Java does:
       // when checking subtyping/type equality, enter constraints
       // derived from the existentially quantified type into the typing environment
       // (aka \Gamma, which tracks types for variables and constraints/kinds for types)
       // as a nice bonus, delaying this until we need it avoids cyclic errors
-      if (_sharpenQuantifierBounds) sharpenQuantifierBounds
+      @inline def tpars                 = underlying.typeSymbol.initialize.typeParams
+      // Is this existential of the form: T[Q1, ..., QN] forSome { type Q1 >: L1 <: U1, ..., QN >: LN <: UN}
+      @inline def isStraightApplication = (quantified corresponds underlying.typeArgs){ (q, a) => q.tpe =:= a }
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala]: I'm dubious the inlining these is worthwhile.

File: src/reflect/scala/reflect/internal/Types.scala [hunk 3]
```diff
@@ -2564,43 +2568,39 @@ trait Types
...
-              quant setInfo tparam.info.substSym(tpars, quantified)
-          }
-        }
-      }
-
-      _sharpenQuantifierBounds = false
-    }
-    private[this] var _sharpenQuantifierBounds = true
+    override def skolemizeExistential(owner0: Symbol, origin: AnyRef) = {
+      val owner                 = owner0 orElse quantifierOwner
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala]: The column alignment of `=` is off.

**Comment by retronym** [comment]
Needs a touch more commentary in the code + commit message (to explain what was wrong before and why this approach is safer). Otherwise I think it is good to go.

**Comment by adriaanm** [comment]
thanks, updated

**Comment by retronym** [comment]
LGTM

**Pull request merged**